### PR TITLE
feat: catch precondition errors

### PR DESCRIPTION
**Pull request by y-lakhdar**
## Proposed changes

Previously, since the preconditions where retuning `false` when not respected, the error message was not [handled in the `catch` method](https://oclif.io/docs/error_handling#error-handling-in-the-catch-method) and, therefore, not tracked in the UA. 

The idea is to throw [PreconditionError](https://github.com/coveo/cli/blob/63cb1223c3308b7dc28ac3bf6a49e8b76a492147/packages/cli/src/lib/errors/preconditionError.ts) whenever a precondition is not respected.

This will facilitate the integration of Amplitude in the CLI when we have to track errors.

Furthermore, having a  [PreconditionError](https://github.com/coveo/cli/blob/63cb1223c3308b7dc28ac3bf6a49e8b76a492147/packages/cli/src/lib/errors/preconditionError.ts) class will make sure errors are consistent in the Amplitude dashboards.

### Commits
* Throw `PreconditionError` instead of returning false (https://github.com/coveo/cli/commit/52fc19992ec15fed529397c2da3047f95eb26a28)
* Simplify some unit tests (https://github.com/coveo/cli/commit/c3a830b85257fd44da65e3542b119594abd72f0f, https://github.com/coveo/cli/commit/aebc606bed5a4a51816cb6b42aa43c0990da971c)
* Remove the `I` from `IDetailedReportable` (https://github.com/coveo/cli/commit/869e05cee5118079beef1b6eb00ca2969a06b55c)

## Testing

- [ ] Unit Tests:

- [ ] Functionnal Tests:

- [ ] Manual Tests:

### CODE REVIEW

File: packages/cli/src/__test__/preconditionUtils.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+const thrower = (reason: string) => {
+  throw new PreconditionError(`${reason} Precondition Error`);
+};
+
+export const mockPreconditions = <Type extends Record<string, boolean>>(
+  preconditionStatus: Type
+) => {
+  type StatusPromises = Record<
+    keyof typeof preconditionStatus,
```
Comment (louis-bompart) [inline packages/cli/src/__test__/preconditionUtils.ts:12]: Can't we use `keyof Type` ?

File: packages/cli/src/__test__/preconditionUtils.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import Command from '@oclif/command';
+import {PreconditionError} from '../lib/errors/preconditionError';
+
+const thrower = (reason: string) => {
+  throw new PreconditionError(`${reason} Precondition Error`);
+};
+
+export const mockPreconditions = <Type extends Record<string, boolean>>(
```
Comment (louis-bompart) [inline packages/cli/src/__test__/preconditionUtils.ts:8]: 'Type' is quite generic for a type name

**Review by louis-bompart — APPROVED** [review]

File: packages/cli/src/lib/decorators/preconditions/preconditions.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import type Command from '@oclif/command';
 export type PreconditionFunction = (
   target: Command,
   instance?: Command
-) => never | void | Promise<never | void>;
+) => Promise<never | void>;
```
Comment (y-lakhdar) [inline packages/cli/src/lib/decorators/preconditions/preconditions.ts:7]: I think it does not make sense to not return a promise if at the end we do
```
await precondition.call(this, target);
```

File: packages/cli/src/__test__/preconditionUtils.ts [hunk 4]
```diff
@@ -1,24 +1,25 @@
...
+  PreconditionStatus extends Record<string, boolean>
+>(
+  preconditionStatus: PreconditionStatus
 ) => {
-  type StatusPromises = Record<
-    keyof typeof preconditionStatus,
-    (_target: Command) => Promise<void>
-  >;
+  type preconditionKeys = keyof typeof preconditionStatus;
+  type preconditionPromises = Record<preconditionKeys, PreconditionFunction>;
```
Comment (y-lakhdar) [inline packages/cli/src/__test__/preconditionUtils.ts:15]: Let's use the type we already have defined in the precondition file

**Review by y-lakhdar — COMMENTED** [review]
@louis-bompart I applied your request and took the liberty to bring some improvements.
Let me know what you think

File: packages/cli/src/lib/decorators/preconditions/preconditions.ts (hunk 3 above)
Comment (louis-bompart) [reply packages/cli/src/lib/decorators/preconditions/preconditions.ts:7]: > I think it does not make sense to not return a promise if at the end we do
> 
> ```
> await precondition.call(this, target);
> ```

An async function always returns a Promise.

Comment (y-lakhdar) [reply packages/cli/src/lib/decorators/preconditions/preconditions.ts:7]: yes, that's what I am saying :) 
hence the `never | void` removal

Comment (louis-bompart) [reply packages/cli/src/lib/decorators/preconditions/preconditions.ts:7]: My bad! I misread your comment at first lgtm! 👍

**Pull request merged**